modules/mod-deep-learning/DeepModel.cpp

```diff
@@ -0,0 +1,199 @@
+   Hugo Flores Garcia
+
+******************************************************************/
+
+#include "DeepModel.h"
+
+#include <torch/script.h>
+#include <torch/torch.h>
+
+#include "FileNames.h"
+#include "WaveTrack.h"
+#include "WaveClip.h"
+
+#include "CodeConversions.h"
+
+// DeepModel Implementation
+
+FilePath DeepModel::DLModelsDir()
+{
+   return FileNames::MkDir( wxFileName( FileNames::DataDir(), wxT("deeplearning-models") ).GetFullPath() );
+}
+
+FilePath DeepModel::BuiltInModulesDir()
+{
+   return FileNames::MkDir( wxFileName( FileNames::BaseDir(), wxT("deeplearning-models") ).GetFullPath() );
+}
+
+void DeepModel::LoadResampler()
+{
+   // load the resampler module
+   std::string resamplerPath = audacity::ToUTF8(wxFileName(BuiltInModulesDir(), wxT("resampler.pt"))
+                                       .GetFullPath());
+   try
+   {
+      mResampler = std::make_unique<torch::jit::script::Module>
+                     (torch::jit::load(resamplerPath, torch::kCPU));
+      mResampler->eval();
+   }
+   catch (const std::exception &e)
+   {
+      Cleanup();
+      throw ModelException(XO("Error an error occurred while loading the resampler"), e.what());
+   }
+}
+
+void DeepModel::Load(const std::string &modelPath)
+{
+   try
+   { 
+      // set mResampler
+      LoadResampler();
+
+      // set mModel
+      mModel = std::make_unique<torch::jit::script::Module>
+                  (torch::jit::load(modelPath, torch::kCPU));
+      mModel->eval();
+      
+      // finally, mark as loaded
+      mLoaded = true;
+   }
+   catch (const std::exception &e)
+   {
+      Cleanup();
+      throw ModelException(XO("Error while loading model"), e.what());
+   }
+}
+
+void DeepModel::Load(std::istream &bytes)
+{
+   try
+   {
+      // load the resampler module
+      LoadResampler();
+
+      // load the model to CPU, as well as the metadata
+      mModel = std::make_unique<torch::jit::script::Module>
+                     (torch::jit::load(bytes, torch::kCPU));
+      mModel->eval();
+      
+      // finally, mark as loaded
+      mLoaded = true;
+   }
+   catch (const std::exception &e)
+   {
+      Cleanup();
+      throw ModelException(XO("Error while loading model"), e.what());
+   }
+}
+
+bool DeepModel::IsLoaded() const
+{
+   return mLoaded; 
+}
+
+void DeepModel::Save(const std::string &modelPath) const
+{
+   if (!mLoaded)
+      throw ModelException(XO("attempted save when no module was loaded."), "");
+
+   mModel->save(modelPath);
+}
+
+void DeepModel::SetCard(ModelCardHolder card)
+{
+   // set the card
+   mCard = card;
+
+   // set sample rate
+   mSampleRate = mCard->sample_rate();
+}
+
+ModelCardHolder DeepModel::GetCard() const
+{
+   return ModelCardHolder(mCard);
+}
+
+void DeepModel::Cleanup()
+{
+   // cleanup
+   mModel.reset();
+   mLoaded = false;
+}
+
+torch::Tensor DeepModel::Resample(const torch::Tensor &waveform, int sampleRateIn, 
+                                  int sampleRateOut) const
```

Comment:
This function is used in two places, the analyzer and the effect.

Should it have a simple early exit path in case the sample rates are equal?

Reply:
5ffa9262fd2870e777e192f70b74b4d64b4e1638